portal-bridge/src/bridge/state.rs

```diff
@@ -253,73 +219,174 @@ impl StateBridge {
-    ) -> anyhow::Result<()> {
-        let timer = metrics.start_process_timer("gossip_state_content");
-        match gossip_state_content(portal_client, content_key.clone(), content_value).await {
-            Ok(_) => {
-                debug!("Successfully gossiped state proof: {content_key}")
-            }
-            Err(msg) => {
-                warn!("Error gossiping state proof: {content_key} {msg}",);
+    async fn acquire_offer_permit(&self) -> OwnedSemaphorePermit {
+        self.offer_semaphore
+            .clone()
+            .acquire_owned()
+            .await
+            .expect("to be able to acquire semaphore")
+    }
+}
+
+// Stats for tracking the performance of offered state content
+struct Stats {
+    keys: HashMap<StateContentKey, OfferReport>,
+}
+
+impl Stats {
+    fn new() -> Self {
+        Self {
+            keys: HashMap::new(),
+        }
+    }
+
+    fn add(&mut self, key: StateContentKey, total: usize) {
+        self.keys.insert(key, OfferReport::new(total));
+    }
+
+    fn update(&mut self, key: &StateContentKey, enr: &Enr, success: bool) {
+        let report = match self.keys.get_mut(key) {
+            Some(report) => report,
+            None => {
+                warn!(
+                    "Expected to find key: {:?}, there's likely a bug in the state stats.",
+                    key
+                );
+                return;
             }
+        };
+        report.update(enr, success);
+        if report.total == report.success.len() + report.fail.len() {
+            report.report(key);
+            // remove the key from the stats
+            self.keys.remove(key);
         }
-        metrics.stop_process_timer(timer);
-        Ok(())
+    }
+}
+
+// Individual report for outcomes of offering a state content key
+struct OfferReport {
+    total: usize,
+    success: Vec<Enr>,
+    // includes both enrs that rejected offer and failed txs
+    // in the future we may want to differentiate between these cases
+    // and return a more detailed report from `trace_offer`
+    fail: Vec<Enr>,
+}
+
+impl OfferReport {
+    fn new(total: usize) -> Self {
+        Self {
+            total,
+            success: Vec::new(),
+            fail: Vec::new(),
+        }
+    }
+
+    fn update(&mut self, enr: &Enr, success: bool) {
+        if success {
+            self.success.push(enr.clone());
+        } else {
+            self.fail.push(enr.clone());
+        }
+    }
+
+    fn report(&self, content_key: &StateContentKey) {
+        info!(
+            "Successfully offered to {}/{} peers. Content key: {}",
+            self.success.len(),
+            self.total,
+            content_key.to_hex(),
+        );
+        debug!(
+            "Successfully offered to {}/{} peers. Content key: {}. Success: {:?}, Fail: {:?}",
+            self.success.len(),
+            self.total,
+            content_key.to_hex(),
+            self.success,
+            self.fail,
+        );
```

Comment:
```suggestion
    fn report(&self, content_key: &StateContentKey) {
        if enabled!(Level::DEBUG) {
                 debug!(
                    "Successfully offered to {}/{} peers. Content key: {}. Success: {:?}, Fail: {:?}",
                    self.success.len(),
                    self.total,
                    content_key.to_hex(),
                    self.success,
                    self.fail,
                );     
        } else {
                  info!(
                      "Successfully offered to {}/{} peers. Content key: {}",
                      self.success.len(),
                      self.total,
                      content_key.to_hex(),
                  );
        }
```

Doing it this way you get the same result, but without the double logging of the same information

Reply:
+1 for Kolby's idea.

But I would only log failed ones at the debug level.
If you think it's worth to log the enrs of successful ones as well, maybe do it at TRACE level.